### ISSUE TITLE
Not all memory addresses replaced by their symbol names in the Disassembly view

### ISSUE THREAD

**Issue by gilbertfrancois** [openMSX/openMSX#1553]
First of all, hurray for the build-in debugger in openMSX. This is a great feature.

I loaded the symbol file from VASM. The parser was able to extract all symbols, as seen in the **Symbol Manager** and most of the symbols are visible in the **Disassembly** view.  But labeled memory addresses are not replaced by the symbol name in the  **Disassembly** view. See attached screenshot.

<img width="1182" alt="Screenshot 2023-11-15 at 09 16 53" src="https://github.com/openMSX/openMSX/assets/15607023/03f088cc-4984-4e0c-a769-749960f6621d">

**Comment by m9710797** [comment]
Hi gilbertfrancois,

I'm glad you like it.

I remember thinking about this: replacing values in instructions like `ld hl,nn`, but i don't recall why I didn't add it, perhaps I just forgot about it.

Note that in general it's not always correct to replace values with symbol names (correct as in: the original meaning of the code). For example an instruction like `ld a,12` should probably not be replaced by `ld a,RDSLT`. Similarly it may, or it may not be a good idea to replace `ld hl,12` with `ld hl,RDSLT`. With your suggestion we would (always) do the substitution in the `ld hl,nn` instruction, but not in the `ld a,n` instruction, right?

Wouter

**Comment by gilbertfrancois** [comment]
Good point, and impossible to have a one-size-fits-all solution. I think I follow your proposal and therefore my subjective opinion would be:

- Substitude only values referring to hl, bc, de, as found in the symbol list.
- Don't substitute numbers from register a
- Don't substitute low values, which may or may not be a reference to a bios call.

Alternatively, this could be a user's choice from a check box, but that is more work in the UI.

**Comment by mthuurne** [comment]
What do we know about a symbol? Only name and address, or also whether it belongs to code or data? Loading code symbols into a register pair can happen, but it is rare. I think the only common example is an inter-slot call. So if we have code vs data information, it could be used as part of a heuristic.

**Comment by m9710797** [comment]
@mthuurne 
Unfortunately we only know the name and value about a symbol, nothing more. Most (all?) of the symbol file formats we support also don't contain extra information.

@gilbertfrancois 
My preference would be to keep the rules simple. So for example let the decision to substitute values by labels NOT depend on the numeric value.
So that would mean we (try to) substitute ALL 16-bit values (also when they are less than 256), and NO 8-bit values (also not when we happen to have a symbol with that numeric value).

And maybe it's not too bad if we sometimes do an unwanted substitution: (today already) when you hover the mouse pointer over a substituted name, you get a tooltip with the original numerical value.

An improvement that was already planned (though low priority) is:
When you have multiple symbols with the same numeric value (but different names), then we currently arbitrarily pick one of the names. In the future we want to make it possible to show the alternative names. For example in the tooltip. And/or we can allow to cycle through the different names by clicking on them (just a single global counter for all the substitutions).

**Comment by gilbertfrancois** [comment]
I see that the problem is not simple. For me, it would be nice to have (not a necessity) to see for example
`ld hl, (_variable)`  and `ld hl, _variable` depending of loading the value or the address into the register.  But if that is complicated due to the fact that substitution is not always intended in the code, then I am totally fine with no substitution here. Maybe @grauw and @mthuurne have better opinions on what to choose. 

tldr; No change on this issue is good too :) I know now that it is not a bug.

**Comment by MBilderbeek** [comment]
Related, BadWolf359 on msx.org writes:
> But what would be even better is if openMSX could distinguish between code labels and constants or variables.

Sounds useful?
See also https://www.msx.org/forum/msx-talk/openmsx/the-best-3-tricks-for-openmsx-debugger-so-far

**Comment by m9710797** [comment]
I've implemented this.
Though know that it's only a change in _heuristics_, it may be better for some cases and worse for others (hopefully more frequently better than worse).
